File: jeu/stance.py
"""
MMA Manager - Système de garde, stabilité et dégâts par jambe
Les calf kicks ciblent la jambe AVANT. Changer de garde la met à l'abri,
mais coûte en efficacité si le combattant n'est pas à l'aise des deux côtés.
"""

import random

ORTHODOX = "orthodox"
SOUTHPAW = "southpaw"


class LegDamage:
    """Dégâts suivis jambe par jambe, pas en bloc."""

    def __init__(self):
        self.gauche = 0
        self.droite = 0

    def add(self, cote, amount):
        setattr(self, cote, getattr(self, cote) + amount)

    def total(self):
        return self.gauche + self.droite

    def get(self, cote):
        return getattr(self, cote)


class StanceState:
    """
    Gère la garde du combattant, quelle jambe est exposée,
    et les pénalités liées au changement de garde.
    """

    def __init__(self, garde_naturelle=ORTHODOX, stance_switching=50):
        self.garde_naturelle = garde_naturelle
        self.garde_actuelle = garde_naturelle
        self.stance_switching = stance_switching  # aisance dans la garde inversée
        self.switches = 0                          # nombre de changements dans le combat

    # --- Quelle jambe est devant selon la garde ---
    def jambe_avant(self):
        return "gauche" if self.garde_actuelle == ORTHODOX else "droite"

    def jambe_arriere(self):
        return "droite" if self.garde_actuelle == ORTHODOX else "gauche"

    def en_garde_inversee(self):
        return self.garde_actuelle != self.garde_naturelle

    def switch(self):
        """Change de garde : la jambe abîmée passe derrière."""
        self.garde_actuelle = SOUTHPAW if self.garde_actuelle == ORTHODOX else ORTHODOX
        self.switches += 1

    # --- Pénalité de garde inversée ---
    def penalite_garde(self):
        """
        0 = aucune pénalité (parfaitement ambidextre)
        Jusqu'à ~0.45 pour un combattant qui ne sait pas switcher.
        """
        if not self.en_garde_inversee():
            return 0.0
        return max(0.0, (100 - self.stance_switching) / 100) * 0.45


# =========================================================
#  STABILITÉ
# =========================================================
def stabilite(stance, leg_dmg, base_balance=50):
    """
    La stabilité dépend surtout de l'état de la jambe AVANT (elle porte l'appui
    et le pivot), et un peu de la jambe arrière (elle donne la propulsion).
    Retourne un facteur entre ~0.25 et 1.0
    """
    dmg_avant = leg_dmg.get(stance.jambe_avant())
    dmg_arriere = leg_dmg.get(stance.jambe_arriere())

    perte = (dmg_avant * 1.0 + dmg_arriere * 0.18) / 78
    perte *= (1.4 - base_balance / 200)   # un bon équilibre encaisse mieux
    return max(0.25, 1.0 - min(0.75, perte))


def facteur_puissance(stance, leg_dmg, base_balance=50):
    """
    La puissance vient du sol : jambes cassées = frappes molles.
    La jambe ARRIÈRE compte plus ici (c'est elle qui pousse dans un cross/low kick).
    """
    dmg_arriere = leg_dmg.get(stance.jambe_arriere())
    dmg_avant = leg_dmg.get(stance.jambe_avant())
    perte = (dmg_arriere * 0.7 + dmg_avant * 0.5) / 88
    base = max(0.4, 1.0 - min(0.6, perte))
    return base * (1 - stance.penalite_garde() * 0.6)


def facteur_esquive(stance, leg_dmg, base_balance=50):
    """Se déplacer demande des jambes valides. Footwork mort = cible fixe."""
    return stabilite(stance, leg_dmg, base_balance) ** 1.2


def facteur_precision(stance):
    """En garde inversée, la précision chute si le combattant n'est pas à l'aise."""
    return 1 - stance.penalite_garde()


# =========================================================
#  DÉCISION DE CHANGER DE GARDE
# =========================================================
def veut_switcher(stance, leg_dmg, fight_iq=50, seuil=25):
    """
    Le combattant décide de switcher pour protéger la jambe avant abîmée.
    Un bon switcher ALTERNE en permanence pour répartir les dégâts :
    c'est tout l'intérêt de savoir combattre des deux côtés.
    """
    dmg_avant = leg_dmg.get(stance.jambe_avant())
    dmg_arriere = leg_dmg.get(stance.jambe_arriere())

    # Rien à protéger encore
    if dmg_avant < seuil:
        return False, None

    # L'écart doit justifier le changement : on ne switche pas pour exposer
    # une jambe dans le même état (voire pire).
    ecart = dmg_avant - dmg_arriere
    if ecart < 8:
        return False, None

    # Bénéfice croissant avec l'écart entre les deux jambes
    benefice = ecart / 60

    # Coût : la pénalité qu'on subira APRÈS le switch
    cout = 0.0 if stance.en_garde_inversee() else (100 - stance.stance_switching) / 100 * 0.45

    proba = 0.20 + fight_iq / 200 + benefice - cout
    # Un combattant qui n'a jamais travaillé la garde inversée reste bloqué :
    # il encaisse plutôt que de se retrouver dans une position où il ne sait rien faire.
    if stance.stance_switching < 40:
        proba *= 0.08
    elif stance.stance_switching < 60:
        proba *= 0.45

    if random.random() < max(0.0, min(0.9, proba)):
        cote = stance.jambe_avant()
        return True, f"jambe {cote} à {dmg_avant} (autre à {dmg_arriere})"
    return False, None


# =========================================================
#  CIBLAGE DES KICKS
# =========================================================
def cible_kick(attaquant_stance, defenseur_stance, type_kick):
    """
    Quelle jambe est touchée.
    Le calf kick et le low kick visent naturellement la jambe avant exposée.
    """
    if type_kick in ("calf_kick", "low_kick"):
        # 80% du temps la jambe avant, parfois l'intérieur de cuisse arrière
        if random.random() < 0.8:
            return defenseur_stance.jambe_avant()
        return defenseur_stance.jambe_arriere()
    return defenseur_stance.jambe_avant()


def resolve_leg_kick(attaquant, defenseur, atk_stance, def_stance,
                     atk_legs, def_legs, type_kick, check_stat, kicking_stat):
    """
    Résout un kick bas avec ciblage de jambe et conséquences sur la stabilité.
    """
    puissance = facteur_puissance(atk_stance, atk_legs)
    precision = facteur_precision(atk_stance)
    esquive_def = facteur_esquive(def_stance, def_legs)

    chance = (45 + (kicking_stat - check_stat) * 0.8) * precision - esquive_def * 10
    if random.uniform(0, 100) < max(8, min(90, chance)):
        cote = cible_kick(atk_stance, def_stance, type_kick)
        dmg = int(random.randint(4, 10) * puissance)
        def_legs.add(cote, dmg)
        return "touché", dmg, cote

    # Check réussi : c'est l'attaquant qui prend
    if random.random() < 0.3:
        cote_atk = atk_stance.jambe_arriere()  # on kicke de la jambe arrière
        recul = random.randint(2, 6)
        atk_legs.add(cote_atk, recul)
        return "checké", recul, f"{cote_atk} (retour)"

    return "évité", 0, None

File: jeu/js/stance.js
/**
 * stance.js — portage de stance.py.
 *
 * Garde, stabilite, degats par jambe. Les calf kicks visent la jambe AVANT ;
 * changer de garde la met a l'abri mais coute en efficacite si le combattant
 * n'est pas a l'aise des deux cotes.
 *
 * REGLE DE PORTAGE : on reproduit le comportement de Python, pas son style.
 * Deux pieges concrets ici :
 *   - int() en Python TRONQUE vers zero, ce n'est PAS Math.round ni
 *     Math.floor (identiques seulement sur les positifs). On utilise
 *     Math.trunc partout ou Python ecrit int().
 *   - l'ORDRE des appels au RNG est du code metier, pas du detail : chaque
 *     random() consomme deux mots de 32 bits. Reordonner deux lignes suffit
 *     a faire diverger tout le combat.
 */

const { alea } = require("./alea.js");

const ORTHODOX = "orthodox";
const SOUTHPAW = "southpaw";

class LegDamage {
  constructor() { this.gauche = 0; this.droite = 0; }
  add(cote, montant) { this[cote] += montant; }
  total() { return this.gauche + this.droite; }
  get(cote) { return this[cote]; }
}

class StanceState {
  constructor(gardeNaturelle = ORTHODOX, stanceSwitching = 50) {
    this.garde_naturelle = gardeNaturelle;
    this.garde_actuelle = gardeNaturelle;
    this.stance_switching = stanceSwitching;
    this.switches = 0;
  }
  jambe_avant()  { return this.garde_actuelle === ORTHODOX ? "gauche" : "droite"; }
  jambe_arriere(){ return this.garde_actuelle === ORTHODOX ? "droite" : "gauche"; }
  en_garde_inversee() { return this.garde_actuelle !== this.garde_naturelle; }
  switch() {
    this.garde_actuelle = this.garde_actuelle === ORTHODOX ? SOUTHPAW : ORTHODOX;
    this.switches += 1;
  }
  penalite_garde() {
    if (!this.en_garde_inversee()) return 0.0;
    return Math.max(0.0, (100 - this.stance_switching) / 100) * 0.45;
  }
}

// ------------------------------------------------------------- STABILITE
function stabilite(stance, legDmg, baseBalance = 50) {
  const dmgAvant = legDmg.get(stance.jambe_avant());
  const dmgArriere = legDmg.get(stance.jambe_arriere());
  let perte = (dmgAvant * 1.0 + dmgArriere * 0.18) / 78;
  perte *= (1.4 - baseBalance / 200);
  return Math.max(0.25, 1.0 - Math.min(0.75, perte));
}

function facteur_puissance(stance, legDmg, baseBalance = 50) {
  const dmgArriere = legDmg.get(stance.jambe_arriere());
  const dmgAvant = legDmg.get(stance.jambe_avant());
  const perte = (dmgArriere * 0.7 + dmgAvant * 0.5) / 88;
  const base = Math.max(0.4, 1.0 - Math.min(0.6, perte));
  return base * (1 - stance.penalite_garde() * 0.6);
}

function facteur_esquive(stance, legDmg, baseBalance = 50) {
  return Math.pow(stabilite(stance, legDmg, baseBalance), 1.2);
}

function facteur_precision(stance) { return 1 - stance.penalite_garde(); }

// ------------------------------------------------ DECISION DE SWITCHER
function veut_switcher(stance, legDmg, fightIq = 50, seuil = 25) {
  const dmgAvant = legDmg.get(stance.jambe_avant());
  const dmgArriere = legDmg.get(stance.jambe_arriere());

  if (dmgAvant < seuil) return [false, null];
  const ecart = dmgAvant - dmgArriere;
  if (ecart < 8) return [false, null];

  const benefice = ecart / 60;
  const cout = stance.en_garde_inversee()
    ? 0.0 : (100 - stance.stance_switching) / 100 * 0.45;

  let proba = 0.20 + fightIq / 200 + benefice - cout;
  if (stance.stance_switching < 40) proba *= 0.08;
  else if (stance.stance_switching < 60) proba *= 0.45;

  // /!\ Le random() est tire APRES les deux sorties anticipees ci-dessus.
  // Le remonter en haut de fonction consommerait du hasard sur des appels
  // qui n'en consomment pas cote Python.
  if (alea.random() < Math.max(0.0, Math.min(0.9, proba))) {
    const cote = stance.jambe_avant();
    return [true, `jambe ${cote} à ${dmgAvant} (autre à ${dmgArriere})`];
  }
  return [false, null];
}

// ---------------------------------------------------- CIBLAGE DES KICKS
function cible_kick(attaquantStance, defenseurStance, typeKick) {
  if (typeKick === "calf_kick" || typeKick === "low_kick") {
    if (alea.random() < 0.8) return defenseurStance.jambe_avant();
    return defenseurStance.jambe_arriere();
  }
  return defenseurStance.jambe_avant();
}

function resolve_leg_kick(attaquant, defenseur, atkStance, defStance,
                          atkLegs, defLegs, typeKick, checkStat, kickingStat) {
  const puissance = facteur_puissance(atkStance, atkLegs);
  const precision = facteur_precision(atkStance);
  const esquiveDef = facteur_esquive(defStance, defLegs);

  const chance = (45 + (kickingStat - checkStat) * 0.8) * precision - esquiveDef * 10;
  if (alea.uniform(0, 100) < Math.max(8, Math.min(90, chance))) {
    const cote = cible_kick(atkStance, defStance, typeKick);
    const dmg = Math.trunc(alea.randint(4, 10) * puissance);
    defLegs.add(cote, dmg);
    return ["touché", dmg, cote];
  }

  if (alea.random() < 0.3) {
    const coteAtk = atkStance.jambe_arriere();
    const recul = alea.randint(2, 6);
    atkLegs.add(coteAtk, recul);
    return ["checké", recul, `${coteAtk} (retour)`];
  }

  return ["évité", 0, null];
}

module.exports = {
  ORTHODOX, SOUTHPAW, LegDamage, StanceState,
  stabilite, facteur_puissance, facteur_esquive, facteur_precision,
  veut_switcher, cible_kick, resolve_leg_kick,
};
